### DISCUSSION TITLE
Threshold by Color

### DISCUSSION THREAD

**Discussion by muddywellyboots** [pyvista/pyvista#2347] [Q&A] [answered]
Hi,
I'm trying to threshold an RGB colored mesh by specific colors. Is this possible?
I've been using the threshold filter, but it seems to only work on a 'RGB-normed value.
Here is information on my mesh. 
![image](https://user-images.githubusercontent.com/24653493/158221643-7753691b-082e-481f-b866-089c49f25565.png)

Many thanks.

**Comment by MatthewFlamm** [comment]
Can you give a code example?  I suspect that `RGB-normed` is creating when you plot this mesh.  You should be able to use `RGB` for the threshold, but I _think_ the underlying vtkThreshold class by default thresholds when 'all' components satisfy the criteria.  The ability to change this behavior isn't exposed by PyVista, so this would be a good feature request.

**Reply by MatthewFlamm** [reply]
See https://vtk.org/doc/nightly/html/classvtkThreshold.html#a01722671ba54f5ec62188734d6a43df1 for the options that vtkThreshold class allows here.  The options would be to use a selected component, all components or any component to satisfy the threshold criteria.

**Reply by banesullivan** [reply]
If `vtkThreshold` supports thresholding a multi-component array, we should make sure the `threshold()` filter in PyVista exposes that

**Reply by MatthewFlamm** [reply]
#2349 implements this functionality.

**Reply by muddywellyboots** [reply]
Thank you 👍

**Comment by banesullivan** [comment, marked as answer]
RGB is essentially an `n` by `3` scalar data array on which you'd like to threshold. This is possible, though I'm not immediately sure if the `vtkThreshold` filter used by our `threshold()` method will handle this.

Normally, we threshold based on an `n` by `1` scalar array where which "chop" out any values under `v`. This is conceptually easy to transfer to a 3D space. Think of your RGB values as 3D coordinates in an RGB colorspace. We can threshold in this color space by thresholding on Red, Green, and Blue individually. To simplify this in PyVista, I would create R, G, and B arrays on the mesh and threshold each array sequentially to produce the desired effect:

```py
mesh['R'] = mesh['RGB'][:,0]
mesh['G'] = mesh['RGB'][:,1]
mesh['B'] = mesh['RGB'][:,2]

result = mesh.threshold(0.5, scalars='R').threshold(0.5, scalars='G').threshold(0.5, scalars='B')
```

This is a rather naive, first-pass solution to the problem, but I think it will produce your desired effect.

**Comment by muddywellyboots** [comment]
Hi Both,
Many thanks for your replies, they have solved my problem.

@banesullivan working on the array is perfect for me - ideally  I will have a maximum and minimum value on my threshold and your approach lets me do this easily.

Thanks again.
